Implement aggregated ServiceImport port merging

tpantelis opened the pull request:
Compute the union of the constituent cluster's ports contained in their `EndpointSlices` as per the MCS spec. This is only done for ClusterIP services.

tpantelis committed:
Implement aggregated ServiceImport port merging

Compute the union of the constituent cluster's ports contained in
their EndpointSlices as per the MCS spec. This is only done for
ClusterIP services.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

vthapar commented on pkg/agent/controller/service_import_aggregator.go:
```diff
@@ -57,28 +60,63 @@ func (a *ServiceImportAggregator) updateOnCreateOrUpdate(name, namespace string)
+			mcsv1a1.LabelServiceName:       serviceName,
+		}).String(),
+	})
+	if err != nil {
```
Should we retry on error here, or requeue logic in caller will take care of it?

tpantelis replied:
yes the latter.

vthapar commented on pkg/agent/controller/service_import_aggregator.go:
```diff
@@ -57,28 +60,63 @@ func (a *ServiceImportAggregator) updateOnCreateOrUpdate(name, namespace string)
+	for i := range list.Items {
+		eps := a.converter.toEndpointSlice(&list.Items[i])
+		si.Spec.Ports = slices.Union(si.Spec.Ports, a.converter.toServicePorts(eps.Ports), func(p mcsv1a1.ServicePort) string {
+			return fmt.Sprintf("%s%s%d", p.Name, p.Protocol, p.Port)
```
What about `p.AppProtocol`? It is another field in ServicePort but not seen it used much ouside of service without selectors. We can revisit this if run into any bugs due to services using appProtocol.

tpantelis replied:
yeah I've wondering about that. We don't use it in the plugin - not sure what it's for.

skitt replied:
It’s used to define the application-level protocol used by the service, see https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol — I suspect it doesn’t get used all that often, but it’s supposed to be carried over to `EndpointSlice`s.

tpantelis replied:
It's unclear to me how it's actually used in K8s other than propagating it to the `EndpointSlices`, which we're doing as well.  Would a user ever define two ports with matching `Name`, `Protocol` and `Port` but differing `AppProtocol`? Not sure...here we wouldn't differentiate them but the aggregated ports are only for compliance anyway, ie we don't use them.  But perhaps we should differentiate `AppProtocol` here in case someone else might use them.

Either way, like I said, we don't use `AppProtocol` in the plugin for resolution so it doesn't seem like we should include it for conflict detection and intersection in the plugin.

tpantelis replied:
Oops correction - we're not propagating `AppProtocol` to the `EndpointSlices`. I can change that. 

Do you guys think we should differentiate `AppProtocol` when calculating the aggregated ports?

skitt replied:
I think we should propagate it, but not use it to differentiate — it’s supposed to describe the traffic, not distinguish multiple services on the same protocol/port.

tpantelis replied:
OK - makes sense. I'll push a separate commit to propagate it.

tpantelis committed:
Propagate AppProtocol to EndpointSlice ports

From the K8s docs, "the value of this field is mirrored by
the corresponding Endpoints and EndpointSlice objects".

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>
